radicle-cob/src/change_graph/evaluation.rs

```diff
@@ -0,0 +1,153 @@
+    identities: &'a I,
+    root: Oid,
+    items: impl Iterator<Item = (&'b Change, Vec<Oid>)>,
+) -> history::History<Id>
+where
+    &'a I: identity::Storage<Identifier = Id>,
+    Id: Clone + PartialEq,
+    Id: std::fmt::Debug,
+    K: sign::Key,
+    for<'de> K::Signature: serde::Deserialize<'de>,
+{
+    let entries = pruning_fold::pruning_fold(
+        HashMap::new(),
+        items.map(|(change, children)| ChangeWithChildren {
+            change,
+            child_commits: children,
+        }),
+        |mut entries, c| match evaluate_change::<_, K>(identities, c.change, &c.child_commits) {
+            Err(reason) => {
+                reason.log(c.change);
+                ControlFlow::Break(entries)
+            }
+            Ok(entry) => {
+                tracing::trace!(commit=?c.change.id(), "change accepted");
+                entries.insert((*c.change.id()).into(), entry);
+                ControlFlow::Continue(entries)
+            }
+        },
+    );
+    // SAFETY: The caller must guarantee that `root` is in `items`
+    history::History::new(root, entries).unwrap()
+}
+
+fn evaluate_change<'a, I, K>(
+    identities: &'a I,
+    change: &Change,
+    child_commits: &[Oid],
+) -> Result<history::Entry<<&'a I as identity::Storage>::Identifier>, RejectionReason>
+where
+    &'a I: identity::Storage,
+    <&'a I as identity::Storage>::Identifier: PartialEq,
+    K: sign::Key,
+    for<'de> K::Signature: serde::Deserialize<'de>,
+{
+    // Check the change signatures are valid
+    {
+        match change.valid_signatures::<K>() {
+            Ok(true) => {}
+            Ok(false) => return Err(RejectionReason::InvalidSignatures),
+            Err(err) => {
+                tracing::warn!(err=%err, "failed to parse signatures");
+                return Err(RejectionReason::InvalidSignatures);
+            }
+        };
+    }
+
+    let author = identities
+        .find_by_oid(*change.author())
+        .map_err(|e| RejectionReason::ErrorFindingAuthor {
+            author_commit_oid: *change.author(),
+            error: Box::new(e),
+        })?
+        .ok_or_else(|| RejectionReason::MissingAuthor {
+            missing_author_oid: *change.author(),
+        })?;
+
+    Ok(history::Entry::new(
+        *change.id(),
+        author.identifier(),
+        change.authority,
+        child_commits.iter().cloned(),
+        change.contents().clone(),
+    ))
+}
+
+struct ChangeWithChildren<'a> {
+    change: &'a Change,
+    child_commits: Vec<Oid>,
+}
+
+impl<'a> pruning_fold::GraphNode for ChangeWithChildren<'a> {
+    type Id = Oid;
+
+    fn id(&self) -> &Self::Id {
+        self.change.id()
+    }
+
+    fn child_ids(&self) -> &[Self::Id] {
+        &self.child_commits
+    }
+}
+
+enum RejectionReason {
+    InvalidSignatures,
+    MissingAuthor {
+        missing_author_oid: Oid,
+    },
+    ErrorFindingAuthor {
+        author_commit_oid: Oid,
+        error: Box<dyn std::error::Error>,
+    },
```

Comment:
Maybe these author-related errors aren't needed anymore if we're just verifying a signature with a key?

Reply:
So I think we still need to _find_ a key. Which is something we need to think about. At the moment we're not content addressing the `did:key`, right? But that probably won't be the case in the future if we need a history of the keys.

Reply:
So `Change` has fields for `author` and `authorizing` which are the content identifiers for the identities of who made the change and the parent identity. So those need to be looked up as commits in the internal logic.

Perhaps you're suggesting that the `author` should be removed? But that doesn't really leave room for tagging the change with metadata.

Reply:
Oh well there's a `signatures` field.

> author: author information; keep this, but it's not used for validation

I'll look into this

Reply:
Ok, if the signatures field includes the pubkey, that should work.

Reply:
Hmmmm so I think what you're looking for is that the author and authorizing should be optional, would that be right? These rejections would go away if it was ok to not find an author in the git backend.

Reply:
Well, the author is optional because it's not needed for verification. It should just be a `did` or something.
The authorizing identity is the project under which the cob lives, so I think that's probably needed?

Reply:
>  The authorizing identity is the project under which the cob lives, so I think that's probably needed?

So before we were carrying it to do the authorization logic within this library. But I've shifted the concerns to the end user -- i.e. the authority's content id is stored in the history and they can choose to fetch that identity and do validation in the `traverse`. So it's not necessarily needed, but we can still make it mandatory.

Reply:
Ah gotcha, but this sounds like it's redundant? I would think that only one authorizing identity is needed per change graph?

Reply:
The identity can change over time, so you need to keep the current snapshot of that identity -- for checking what keys are in that snapshot, for example.

Reply:
Ah, so not the id, but the commit hash of the 'current' doc?

Reply:
Ya, exactly :) If you look at the tests, I'm resolving the `authorizing` to an `Oid` to look up that doc and filtering by the author name. The key use case would be similar but filtering on `is_delegate` of some sort.

Reply:
I was able to convince myself that, in fact, I'm wrong. `authorizing` being the project does need to be passed because the reference it's stored under needs its identifier. I still think it makes sense to set it in the `Entry` for inspecting the project at that `Oid` though.

However, perhaps what I should have explored is the `author` being optional, or at least the `PublicKey` and then the optional "identity" (DID or otherwise) -- which is what I think you suggested in discord :)